stan-dev/math
# fix threading for clang, fixes #844

#### Submission Checklist


#### Summary:
clang + threading is broken right now. This PR adds testing on travis with clang and threading along with the fixes to the AD stack to make it work again. To see things failing, please use `develop` for this.

```
[21:01:42][sebi@sebastians-macbook-pro-1:~/work/performance-tests-cmdstan]$ ./comparePerformance.py performance.csv develop_performance.csv
('stat_comp_benchmarks/benchmarks/gp_pois_regr/gp_pois_regr.stan', 0.99)
('stat_comp_benchmarks/benchmarks/low_dim_corr_gauss/low_dim_corr_gauss.stan', 1.04)
('stat_comp_benchmarks/benchmarks/eight_schools/eight_schools.stan', 1.03)
('stat_comp_benchmarks/benchmarks/irt_2pl/irt_2pl.stan', 0.98)
('compilation', 1.09)
('stat_comp_benchmarks/benchmarks/gp_regr/gp_regr.stan', 0.99)
('stat_comp_benchmarks/benchmarks/low_dim_gauss_mix_collapse/low_dim_gauss_mix_collapse.stan', 0.94)
('stat_comp_benchmarks/benchmarks/pkpd/one_comp_mm_elim_abs.stan', 0.99)
('stat_comp_benchmarks/benchmarks/sir/sir.stan', 0.97)
('stat_comp_benchmarks/benchmarks/pkpd/sim_one_comp_mm_elim_abs.stan', 0.96)
('stat_comp_benchmarks/benchmarks/garch/garch.stan', 1.38)
('stat_comp_benchmarks/benchmarks/gp_regr/gen_gp_data.stan', 0.89)
('stat_comp_benchmarks/benchmarks/arK/arK.stan', 1.04)
('stat_comp_benchmarks/benchmarks/arma/arma.stan', 1.04)
('stat_comp_benchmarks/benchmarks/low_dim_gauss_mix/low_dim_gauss_mix.stan', 0.92)
```

#### Intended Effect:
Add test coverage for clang+threading and fix it.

#### How to Verify:
1. enable threading with `CXXFLAGS+=-DSTAN_THREADS` in your `make/local`
2. use `CC=clang++`
3. you can confirm that this is broken on `develop` and works on this branch by running the test `test/unit/math/rev/mat/functor/gradient_test.cpp`

#### Side Effects:
Access to the AD stack happens by default through the instance() function of the stack. However, for performance reasons (as doced) the access in the `vari` class happens through function  local static references (which are thread_local if needed).

#### Documentation:
Additional comments added where appropriate.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
